### PR TITLE
 feat(protocol+broker): add user task record

### LINKED ISSUE #14996: Create a Zeebe User Task record and event structure
**Is your feature request related to a problem? Please describe.**

I want to have User Tasks as first-class citizens in Zeebe, represented on the log stream.
There is no specific record or event structure in Zeebe to do this.

**Describe the solution you'd like**

```[tasklist]
### Tasks
- [x] There is a basic record and event structure for User Tasks that can be stored on the log stream.
- [x] The record and events represent the values a User Task can have.
- [x] The record is supported by our exporters (ES and OS index templates are provided).
```

**Describe alternatives you've considered**

Keep using Job records to represent User Tasks on the log stream.

**Additional context**

* Base requirement to enable https://github.com/camunda/zeebe/issues/14938.
* Guide for [adding new records in Zeebe](https://github.com/camunda/zeebe/blob/main/docs/developer_handbook.md#how-to-create-a-new-record).

### PR DESCRIPTION
**Pull request by tmetzke**
## Description

* Adds the record and intent structure for user tasks

## Related issues

closes #14996

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by tmetzke** [comment]
@koevskinikola, I am not entirely sure about the `variables` attribute in the user task record. To be honest, I copied this from the job record. 

The variables are never exported for jobs (as far as I can see in the ES and OS templates) but are part of the record classes. I assume that it's for transport reasons, to allow to pass on variables in the gateway communication (to and from clients)? Having this on user task records might come in handy for user task listeners later. We'll deal with user task records there (I presume) that should also bring along variables then but never export them in ES or OS.

Let me know what you think 🙂

**Comment by koevskinikola** [comment]
Hey @tmetzke,

>I am not entirely sure about the variables attribute in the user task record. To be honest, I copied this from the job record.
>
> The variables are never exported for jobs (as far as I can see in the ES and OS templates) but are part of the record classes. I assume that it's for transport reasons, to allow to pass on variables in the gateway communication (to and from clients)? Having this on user task records might come in handy for user task listeners later. We'll deal with user task records there (I presume) that should also bring along variables then but never export them in ES or OS.

Yes, in `JobRecords` we have variables so we can transport them between the Gateway and Broker when activating/completing jobs. We don't keep them in the `JobState` but just persist them and fetch them from the `VariableState`. This avoids duplication in the state and when exporting.

They might come in useful for user task listeners, maybe to use the `UserTaskRecord` as a way to fetch the variables that need to be included in the listener jobs.

I think they will definitely come in useful when completing a User Task. From what I remember, it was possible to set variables explicitly when completing a User Task in C7.

**Review by koevskinikola — CHANGES_REQUESTED** [review]
The changes look great @tmetzke.

:wrench: I have some suggestions and an open question. You can see them below.

:wrench: We should also adjust the `EnumSet` range [here](https://github.com/camunda/zeebe/blob/fd8ce99cc2e01fb3186a45f33605c5a39ecd2dbe/engine/src/main/java/io/camunda/zeebe/engine/Engine.java#L48-L49) to include `ValueType.USER_TASK` (and indirectly `FORM`). Otherwise, User Task commands won't be processed by the engine. We can also do this later if you like, but we might forget.

File: protocol/src/main/java/io/camunda/zeebe/protocol/record/intent/UserTaskIntent.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @Override
+  public short value() {
+    return value;
+  }
+
+  @Override
+  public boolean shouldBanInstanceOnError() {
+    return false;
```
[the remark this answers is not in the document]
Comment (tmetzke) [reply protocol/src/main/java/io/camunda/zeebe/protocol/record/intent/UserTaskIntent.java:67]: Good point 👍 I'll adjust that to behave the same way the job intents do 🙂

File: protocol/src/main/java/io/camunda/zeebe/protocol/record/value/UserTaskRecordValue.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  String getCandidateUsers();
+
+  String getDueDate();
+
+  String getFollowUpDate();
+
+  long getFormKey();
+
+  /**
+   * @return the element id of the corresponding service task
```
[the remark this answers is not in the document]
Comment (tmetzke) [reply protocol/src/main/java/io/camunda/zeebe/protocol/record/value/UserTaskRecordValue.java:48]: Thanks, good catch! 👍
It's funny, I looked through the code twice to find copy-paste leftovers but apparently, at some point, you get copy-paste-blind or something like that 😄

**Comment by tmetzke** [comment]
> 🔧 We should also adjust the `EnumSet` range [here](https://github.com/camunda/zeebe/blob/fd8ce99cc2e01fb3186a45f33605c5a39ecd2dbe/engine/src/main/java/io/camunda/zeebe/engine/Engine.java#L48-L49) to include `ValueType.USER_TASK` (and indirectly `FORM`). Otherwise, User Task commands won't be processed by the engine. We can also do this later if you like, but we might forget.

I've added a task to **Increment 2** in the [EPIC](https://github.com/camunda/zeebe/issues/14938) to do this. In short, it will be easier to do this there.

Longer story: With increment 2, we'll actually write events to the log stream by processing a model accordingly and can directly add related tests that verify the replay. Otherwise, we'd either just have to add this here and test it there or add an event manually to the log stream here in the replay test (which needs additional test capabilities we'd have to add here, too).

Let me know what you think.

**Comment by tmetzke** [comment]
@koevskinikola, please don't merge directly after review (if successful), I want to squash one of the commits 🙂

**Review by koevskinikola — APPROVED** [review]
Looks great @tmetzke :+1: 

I agree with your comment [here](https://github.com/camunda/zeebe/pull/15171#issuecomment-1812180040). So feel free to merge. :smile:

**Comment by tmetzke** [comment]
bors merge

**Pull request merged**